### DISCUSSION TITLE
Transaction blocks

### DISCUSSION THREAD

**Discussion by DzigaV** [psycopg/psycopg#28] [Q&A] [answered]
Given this:

> psycopg3 has a behaviour that may result surprising compared to psql: by default, any database operation will start a new transaction. As a consequence, changes made by any cursor of the connection will not be visible until Connection.commit() is called

and this:

> If a database operation fails with an error message such as InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block, it means that a previous operation failed and the database session is in a state of error. You need to call rollback() if you want to keep on using the same connection.

Should we always use transaction blocks?

`with conn.transaction() as tx:`

Or does using the cursor context manager do the same thing?

**Comment by dvarrazzo** [comment, marked as answer]
The cursor context manager doesn't handle a transaction.

If you want to handle the transaction as a block you can use `conn.transaction()`, otherwise you can use `conn.commit()` and `conn.rollback()`.
